pkg/ebpf/c/tracee.bpf.c

```diff
@@ -6289,3 +6313,836 @@ int tc_ingress(struct __sk_buff *skb)
+    if (inode == 0)
+        return 0;
+
+    switch (type) {
+        case BPF_CGROUP_INET_INGRESS:
+        case BPF_CGROUP_INET_EGRESS:
+            break;
+        default:
+            return 0; // wrong attachment type, return fast
+    }
+
+    // save args for kretprobe
+    struct entry entry = {0};
+    entry.args[0] = PT_REGS_PARM1(ctx); // struct sock *sk
+    entry.args[1] = PT_REGS_PARM2(ctx); // struct sk_buff *skb
+
+    // prepare for kretprobe using entrymap
+    u32 pid = data.context.task.host_pid;
+    bpf_map_update_elem(&entrymap, &pid, &entry, BPF_ANY);
+
+    // NOTE: The net_task_context is the event_context's "task_context" when the
+    //       socket was created. This could take tracee to situations which the
+    //       event being submited by the cgroup skb ingress/egress eBPF program
+    //       has outdated information (host_ppid, ppid, comm, cgroup_id, uid,
+    //       mnt_id, pid_id, uts_name). The only possible way to solve this is
+    //       to hook to (*file)->fops() functions so, each time a socket is
+    //       poll'ed(), written(), read(), sendmsg'ed(), the task reading or
+    //       writing to the socket will have its task_info updated with correct
+    //       "should_trace" metadata (and then its net_task_context counter part
+    //       could also be updated).
+    //
+    //       Hooking into a socket file's file operations handlers might be
+    //       expensive in terms of performance, and it is fair to assume long
+    //       lasting sockets would be traced with old task context (from the
+    //       time the socket was created). So, for now, this is A DECISION and
+    //       not a lack of support.
+    //
+    //       When a task has an opened socket (or "long lasting" socket), and it
+    //       is migrated to a container that is currently being traced, for
+    //       example, whenever its "task_info" is updated, the should_trace
+    //       logic will allow the cgroup skb egress (not ingress, since it runs
+    //       in a kernel thread context) ebpf programs to start tracing those
+    //       sockets. This is the same technique used for existing sockets of
+    //       tasks that does not have their "task_info" updated.
+
+    // pick original task from socket inode
+    net_task_context_t *netctx = bpf_map_lookup_elem(&inodemap, &inode);
+    if (!netctx) {
+        // ignore if this is a kernel thread (can't map inode <=> task)
+        struct task_struct *task = data.task;
+        if (BPF_READ(task, flags) & PF_KTHREAD)
+            return 0;
+
+        // if not a traced inode: check if task should be traced ...
+        if (should_trace(&data)) {
+            // ... then update inodemap correlating inode <=> task
+            net_task_context_t netctx = {0};
+            get_net_task_context(&data, &netctx);
+            bpf_map_update_elem(&inodemap, &inode, &netctx, BPF_ANY);
+        }
+        return 0;
```

Comment:
HA, I'm glad you realized this. No, but OLD verifiers don't allow me to continue here by any means (I tried many different approaches). This is the original code:

https://github.com/rafaeldtinoco/drafts/blob/new-approach/drafts.bpf.c#L439

which handles what you're saying... which, in this code, would be as simple as:

```C
        // if not a traced inode: check if task should be traced ...
        if (should_trace(&data)) {
            // ... then update inodemap correlating inode <=> task
            net_task_context_t netctx = {0};
            set_net_task_context(&data, &netctx);
            bpf_map_update_elem(&inodemap, &inode, &netctx, BPF_ANY);
        } else
            return 0;
```

but if I try to do that, the old verifiers complain about:

```
; bpf_map_update_elem(&inodemap, &inode, &netctx, BPF_ANY);
913: (18) r1 = 0xffff8e4c44f81c00
915: (b7) r4 = 0
916: (85) call bpf_map_update_elem#2
917: (b7) r1 = 32
918: (79) r3 = *(u64 *)(r10 -416)
919: (0f) r3 += r1
920: (bf) r1 = r10
; u64 skbts = BPF_READ(skb, tstamp);
921: (07) r1 += -248
922: (b7) r2 = 8
923: (85) call bpf_probe_read_kernel#113
last_idx 923 first_idx 917
regs=4 stack=0 before 922: (b7) r2 = 8
; net_event_context_t neteventctx = {
924: (7b) *(u64 *)(r10 -256) = r8
925: (7b) *(u64 *)(r10 -264) = r8
926: (7b) *(u64 *)(r10 -272) = r8
927: (7b) *(u64 *)(r10 -280) = r8
928: (7b) *(u64 *)(r10 -288) = r8
; __builtin_memcpy(&eventctx->task, &netctx->taskctx, sizeof(task_context_t));
929: (79) r1 = *(u64 *)(r6 +8)
R6 invalid mem access 'inv'
processed 551 insns (limit 1000000) max_states_per_insn 0 total_states 32 peak_states 32 mark_read 30
-- END PROG LOAD LOG --
libbpf: prog 'cgroup_bpf_run_filter_skb': failed to load: -13
libbpf: failed to load object 'embedded-core'
{"level":"fatal","ts":1665436266.6429622,"msg":"app","error":"error initializing Tracee: failed to load BPF object: permission denied"}
```

I'm going to add a note here saying it needs some work to allow continuing the same packet. For now it will start picking the very next packet.

Reply:
Hum, can't reproduce the verifier issues I had (it was only happening for clang-12 and older kernels for some reason). Lets see, will keep this change and submit a PR update for test checks.

Reply:
Alright, the code passes now (Verifier was probably upset with something else when I tried this first).